### ISSUE TITLE
Support OIDC without the Kuberentes API Server configuration (something similar to Argo CD configuration)

### ISSUE THREAD

**Issue by zivcex** [headlamp-k8s/headlamp#2614]
## Is your feature request related to a problem? Please describe the impact that the lack of the feature requested is creating.

There are multiple issues with Keycloak OIDC 
https://github.com/headlamp-k8s/headlamp/issues/1651 
https://github.com/headlamp-k8s/headlamp/issues/1436

Even though the OIDC is configured the login doesn't work as expected.

On first login I get the following issue with 401 and going from here headlamp requests to re-login: 
![image](https://github.com/user-attachments/assets/fee7f587-121c-4cc7-b89a-a0b090f1a5fe)
Next up are just 403s with no token attached
![image](https://github.com/user-attachments/assets/82447cd1-883e-48a0-9f3b-dc2630ce86bb)

## Describe the solution you'd like
Process similar to how Argo CD/WF handles oidc : https://argo-workflows.readthedocs.io/en/latest/argo-server-sso/
There's no need for any additional configuration with the steps above (Kubernetes API Server configuration).

## What users will benefit from this feature?
Anyone who wants to configure SSO with RBAC.

## Are you able to implement this feature?
No.

**Comment by joaquimrocha** [comment]
cc/ @knrt10

**Comment by zivcex** [comment]
Hi @joaquimrocha is this something we can expect within the next release? I don't really know how much effort is needed here so that's why I'm asking..

**Comment by joaquimrocha** [comment]
I am not sure either. Sometimes these OIDC issues are tricky. @knrt10 is away today but should be able to check this later this week.

**Comment by davepgreene** [comment]
Seconding this request. Tying auth to the cluster when dealing with an external provider feels kind of odd: like you're using one source (the identity provider) for authentication but using the other (cluster OIDC -> IdP) for authorization rather than just using the IdP for both. In general with a tool like this I'd expect some kind of role-mapping functionality so that user roles in the IdP map to capabilities within Headlamp.

My use case is an in-cluster IdP (Authentik in this case) rather than externally hosted.

**Comment by iSE-stefan-may** [comment]
Running also into this issue. I've implemented a PR for exactly this feature: #2643 
Tested locally and works fine. The helmchart already contains a ServiceAccount with the required permissions.

**Comment by joaquimrocha** [comment]
Thanks @iSE-stefan-may . I have assigned @yolossn and @ashu8912 to this issue as we want to prioritize it. We'll review the PR.

**Comment by iSE-stefan-may** [comment]
Sorry, I made a mistake. Initially, my comment didn't mention my PR but another issue, is now fixed. My PR is #2814

**Comment by yolossn** [comment]
> Seconding this request. Tying auth to the cluster when dealing with an external provider feels kind of odd: like you're using one source (the identity provider) for authentication but using the other (cluster OIDC -> IdP) for authorization rather than just directly using the IdP for both. In general with a tool like this I'd expect some kind of role-mapping functionality so that user roles in the IdP map to capabilities within Headlamp.

 Headlamp doesn't handle any kind of authentication or authorization, It just initiates the login flow if OIDC is configured and the token is forwarded to the Kubernetes API server, which does the authentication and authorization.

**Comment by joaquimrocha** [comment]
@iSE-stefan-may , would something like @yolossn is suggesting help your use-case?

**Comment by iSE-stefan-may** [comment]
@yolossn  @joaquimrocha I'm currently looking into it, but it does not seem to work.

This issue is about when the kubernetes API server does not understand oidc, e.g. if it was setup by another party out of our control. The suggested oauth2-proxy would still only send an oidc token to kubernetes, which it still not understands.

Only proxy which replaces OIDC with the serviceaccount token is kube-oidc-proxy, but it only communicates with the kubernetes API, so it can't be put before headlamp. Putting it between headlamp and kuberneteas API does not work either, because if we use the in-cluster flag we can't override the kubeconfig.

If you have any other solution (e.g. with another proxy) I would try it out it.

**Comment by iSE-stefan-may** [comment]
@yolossn 

I don't understand, how this solves any problem here? The issue is not, that we cannot send OIDC tokens to headlamp, but that if the kubernetes API server is not configured for OIDC headlamp is not authenticated/authorized against the kubernetes API.

So the current situation is:
- headlamp has the OIDC workflow implemented
- headlamp does not evaluate the OIDC tokens besides the OIDC refresh workflow
- headlamp sends the OIDC token to the kubernetes API
- kubernetes evaluates the OIDC token and authenticates the user by it
- if kubernetes does not understand OIDC tokens, OIDC token forwarding in headlamp is useless

**Comment by illume** [comment]
@zivcex, @davepgreene I'm interested in what you think of the PR by @iSE-stefan-may ? Does this solve things for you? / Is this something you would use?


- PR https://github.com/headlamp-k8s/headlamp/pull/2814
- more explanation https://github.com/headlamp-k8s/headlamp/issues/2614#issuecomment-2651432929
- more explanation https://github.com/headlamp-k8s/headlamp/issues/2614#issuecomment-2653199802

Also interested in what you think about the post by @yolossn here:
- https://github.com/headlamp-k8s/headlamp/issues/2614#issuecomment-2651517957

**Comment by zivcex** [comment]
Yeah, I already have a setup that I'd use this for. Mainly for for Keycloak used as IDP for internal tooling. Hence the ArgoCD in the title, as I'm using it in this internal type of setting. The proxy in front of headlamp or kubernetes API server setting just complicates things for me.

**Comment by yolossn** [comment]
> I don't understand, how this solves any problem here? The issue is not, that we cannot send OIDC tokens to headlamp, but that if the kubernetes API server is not configured for OIDC headlamp is not authenticated/authorized against the kubernetes API.

The OAuth2 Proxy manages the OIDC login flow. Upon successful authentication, it forwards requests to the Headlamp service within the cluster. This setup allows Headlamp to operate without the Kubernetes API server being configured for OIDC. Instead, authentication is handled externally by the OAuth2 Proxy, which then proxies authenticated requests to Headlamp. IMHO we should try to solve this outside of headlamp as much as possible.

**Comment by yoadey** [comment]
> The OAuth2 Proxy manages the OIDC login flow. Upon successful authentication, it forwards requests to the Headlamp service within the cluster. This setup allows Headlamp to operate without the Kubernetes API server being configured for OIDC. Instead, authentication is handled externally by the OAuth2 Proxy, which then proxies authenticated requests to Headlamp. IMHO we should try to solve this outside of headlamp as much as possible.

Assuming this case: currently, headlamp does not extract any authentication information from any header given by an oidc proxy. How does headlamp know, what the user is allowed to see/do and what not?
Also, how should headlamp authenticate to the kubernetes API when running in kubernetes itself?

**Comment by iSE-stefan-may** [comment]
> The OAuth2 Proxy manages the OIDC login flow. Upon successful authentication, it forwards requests to the Headlamp service within the cluster. This setup allows Headlamp to operate without the Kubernetes API server being configured for OIDC. Instead, authentication is handled externally by the OAuth2 Proxy, which then proxies authenticated requests to Headlamp. IMHO we should try to solve this outside of headlamp as much as possible.

Assuming this case: currently, headlamp does not extract any authentication information from any header given by an oidc proxy. How does headlamp know, what the user is allowed to see/do and what not?
Also, how should headlamp authenticate to the kubernetes API when running in kubernetes itself?

**Comment by yolossn** [comment]
> How does headlamp know, what the user is allowed to see/do and what not?
Also, how should headlamp authenticate to the kubernetes API when running in kubernetes itself?

The user can create a service token and use it or the admin can use `injectRequestHeaders` [config](https://oauth2-proxy.github.io/oauth2-proxy/configuration/alpha-config/#alphaoptions) in oauth2-proxy to set the service token in the proxied request, (I am working on setting this up as a POC). If this works as expected, we can add an option in the helm chart to deploy headlamp along with oauth2-proxy configured making it seamless for the admin. Ill update my findings here.

**Comment by zivcex** [comment]
@yolossn have you setup the oauth2-proxy with the above configuration? I've used oauth2-proxy earlier with prometheus and it required a lot of configuration and trial and error..

**Comment by illume** [comment]
I had a number of discussions with people at Kubecon London who were blocked on this issue. 

I asked if something like what Argo CD uses, or a proxy in front would be preferred and the choice was similar to other requests like this:
> The proxy in front of headlamp or kubernetes API server setting just complicates things for me.

@joaquimrocha I think this issue should be prioritized. (for v0.32.0?)

**Comment by joaquimrocha** [comment]
@yolossn / @ashu8912 , you have been looking into OIDC. Anything to report related to this?
Maybe something we should prioritize soon, as Rene suggested.

**Comment by phmcder** [comment]
I'm also hopeful for this. I'm unable to have public facing identity services due to the nature of my work. I'm also unable to make changes to the API service as we use eks and restricted to using docker desktop on our laptops. I'm in the process of testing kube-oauth-proxy but it would be nicer to have a cleaner solution.

**Comment by yolossn** [comment]
Created an issue for supporting headlamp to load service account token that the user can opt-in and configure to make headlamp authenticated by default. Once we have this in place, Ill create a doc for using headlamp with outh2proxy.

https://github.com/kubernetes-sigs/headlamp/issues/3606

**Comment by maplewf** [comment]
> Created an issue for supporting headlamp to load service account token that the user can opt-in and configure to make headlamp authenticated by default. Once we have this in place, Ill create a doc for using headlamp with outh2proxy.
> 
> [#3606](https://github.com/kubernetes-sigs/headlamp/issues/3606)

I don't understand how this workflow implement the RBAC? Does it mean that all users with this workflow will have same permission to login cluster even it's oidc login with oauth2-proxy, right? It did authentication but without authorization.

**Comment by maplewf** [comment]
> [@yolossn](https://github.com/yolossn) [@joaquimrocha](https://github.com/joaquimrocha) I'm currently looking into it, but it does not seem to work.
> 
> This issue is about when the kubernetes API server does not understand oidc, e.g. if it was setup by another party out of our control. The suggested oauth2-proxy would still only send an oidc token to kubernetes, which it still not understands.
> 
> Only proxy which replaces OIDC with the serviceaccount token is kube-oidc-proxy, but it only communicates with the kubernetes API, so it can't be put before headlamp. Putting it between headlamp and kuberneteas API does not work either, because if we use the in-cluster flag we can't override the kubeconfig.
> 
> If you have any other solution (e.g. with another proxy) I would try it out it.

@yolossn If we already setup kube-oidc-proxy in cluster, is it possible for headlamp to allow user to override api endpoint to be kube-oidc-proxy instead of kube api server in kubeconfig? Then it's minimum change for headlamp to support it.

**Comment by sstarcher** [comment]
This is a blocker for us using headlamp.  As it looks like we can't properly use our authentication method due to this flow and interactions between AWS and Azure.

**Comment by DaanSelen** [comment]
> This is a blocker for us using headlamp. As it looks like we can't properly use our authentication method due to this flow and interactions between AWS and Azure.

+1

**Comment by illume** [comment]
@DaanSelen and @sstarcher can you please let us know if using oauth2proxy in front of headlamp won't work for you? It's my understanding that it should be possible to use it for OIDC impersonation.

If it's not possible for you we'd like to know some details, and then we can reconsider adding support into Headlamp itself.

**Comment by sstarcher** [comment]
That is the direction we are headed.  I have not looked, but if we pass through the group claims can those be used in headlamp to give different levels of RBAC?  Or would we be limited to one type of authorization?

**Comment by DaanSelen** [comment]
> [@DaanSelen](https://github.com/DaanSelen) and [@sstarcher](https://github.com/sstarcher) can you please let us know if using oauth2proxy in front of headlamp won't work for you? It's my understanding that it should be possible to use it for OIDC impersonation.
> 
> If it's not possible for you we'd like to know some details, and then we can reconsider adding support into Headlamp itself.

I'll try to put this forward except we (as a company) move quite slow so hang on!

**Comment by illume** [comment]
> That is the direction we are headed. I have not looked, but if we pass through the group claims can those be used in headlamp to give different levels of RBAC? Or would we be limited to one type of authorization?

Yes, you can give users different roles.

**Comment by illume** [comment]
Hey hey.

There's another PR now:

-  #4919 


We need feedback if this would solve your issues.

cc @adamdost-0 @sstarcher @DaanSelen @maplewf @phmcder @zivcex @c58

(The PR has documentation on how to use it, and the PR summarizes the architecture and how it works. It is in the nightly container image to make testing a bit easier until the next release.)

**Comment by c58** [comment]
The feature seems to be working, im using this set of extra args:
```
      extraArgs:
      - -proxy-auth=true
      - -proxy-auth-username-header=Remote-User
      - -proxy-auth-group-header=Remote-Groups
      - -proxy-auth-email-header=Remote-Email
      - -kubeconfig=/home/headlamp/.config/Headlamp/kubeconfigs/config
```

My setup is Traefik -> Authelia -> Headlamp. Im not using any additional atuhorization layer and reuse the service account Headlamp has. This is the kubeconfig i have:
```
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: headlamp-kubeconfig
  namespace: headlamp
data:
  config: |-
    apiVersion: v1
    kind: Config
    current-context: main
    clusters:
      - name: main
        cluster:
          server: https://kubernetes.default.svc
          certificate-authority: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    users:
      - name: main
        user:
          tokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    contexts:
      - name: main
        context:
          cluster: main
          user: main
``` 
Without the config i was asked by headlamp UI to provide service token which i didn't want.
Overall the feature solved my problem, thank you very much to everyone who worked on it! Waiting for the stable release with it to switch from nightly container.

**Comment by illume** [comment]
@c58 thanks for the feedback, and for trying it out.

It should land in the next release 0.43.0 which is scheduled around June 3rd.

**Comment by phmcder** [comment]
> Hey hey.
> 
> There's another PR now:
> 
> * [Proxy auth support #4919](https://github.com/kubernetes-sigs/headlamp/pull/4919)
> 
> We need feedback if this would solve your issues.
> 
> cc [@adamdost-0](https://github.com/adamdost-0) [@sstarcher](https://github.com/sstarcher) [@DaanSelen](https://github.com/DaanSelen) [@maplewf](https://github.com/maplewf) [@phmcder](https://github.com/phmcder) [@zivcex](https://github.com/zivcex) [@c58](https://github.com/c58)
> 
> (The PR has documentation on how to use it, and the PR summarizes the architecture and how it works. It is in the nightly container image to make testing a bit easier until the next release.)

Thanks for asking @illume. Sorry for the long answer. Short answer is this won't help for my specific issue but is a great addition to headlamps capabilities. I would have used it if I had a public oidc provider

For my project we are running inside EKS with AWS managing the authenication with the Kubernetes control plane.  Before I can use kubectl in the cli. I must authenticate myself using `aws sso login`. AWS uses a custom authentication webhook to validate the retuned sigV4 token and map the user within to a K8S RBAC. This obviously is not an authentication flow which Headlamp could use.

The other option is that we can add an alternative OIDC provider to the AWS EKS controller (such as entra) so users can authenticate in a more standard way.  The issue we have here is that we are running a private OIDC provider hosted in our local network. AWS EKS does not support OIDC providers that are not publicly accessible. Therefore we cannot authenticate any token issued by our private OIDC provider with the EKS controller API. 

The only option we have left if to run some K8S token proxy which would receive a request through the Headlamp UI containing an Authorization bearer token issued by our private OIDC provider and then create an internal service account/user token to authenticate the user with the EKS controller.

The change above would help with the first part. Our OIDC works with our proxy to insert the JWT as a new header in each request to Headlamp, but as I said earlier, this token would have been issued by the private OIDC so would be rejected when used to call EKS.

Our work around it to use [Kube-OIDC-Proxy](https://github.com/TremoloSecurity/kube-oidc-proxy/) to act as a proxy to the underlying EKS controller. That does allow tokens created by our provide OIDC provider to be traded for group based impersonation tokens 

The only pain we have with this solution is a user experience one. To access the cluster a user must:

1.  Navigate to headlamp
2. Open the "Home" panel
3. Select the "oidc-proxy" cluster
4. Click "sign in"
5. Authenticate with the private oidc provider
6. Return to headlamp
Ideally we would and to set the oidc-proxy cluster as the default and prompt the user to authenticate as soon as they navigate to headlamp removing steps 1-3
 
Another more simple option should be provided by #3606

**Comment by illume** [comment]
@phmcder thanks for the detailed response.

For:
- https://github.com/kubernetes-sigs/headlamp/issues/3606

There is a PR making way...
- https://github.com/kubernetes-sigs/headlamp/pull/5578

**Comment by illume** [comment]
I think we can close this one now.

This merged PR covers many of the cases mentioned by people:
- #4919 

For the others I think this issue covers it (the PR for that is shaping up nicely too):
- https://github.com/kubernetes-sigs/headlamp/issues/3606

If I'm wrong please feel free to ping us and we can reopen this issue.
